### ISSUE TITLE
speed up boundary_matrix using spatial queries

### ISSUE THREAD

**Issue by jeffreyhanson** [prioritizr/prioritizr#74]
The `boundary_matrix` function calculates the shared boundary length between different planning units. These calculations assume that the vertices in each planning unit are correctly aligned (see #30). To ensure this assumption is met, this function has a pre-processing step which compares every single edge of planning unit to every vertex in every other planning unit to insert missing vertices. This is not terribly efficient. One idea for speeding this process up would be to use GEOS STR Tree's to only compare every single edge of a given planning unit to every single vertex in only the neighbouring planning units. This could potentially speed up this function a lot, but would rely on some experimental functions in _rgeos_ (i.e. `rgeos::gUnarySTRtreeQuery`).

**Comment by ricschuster** [comment]
`rgeos::gUnarySTRtreeQuery` sounds very useful. Maybe add a flag to the `boundary_matrix` function that allows the user to take advantage of `rgeos::gUnarySTRtreeQuery`, with the caveat in the documentation that the function is experimental?

**Comment by jeffreyhanson** [comment]
Yeah, that's a very good point regarding the experimental nature of the function, and I think that's a great idea including the flag and noting it in the docs.

**Comment by jeffreyhanson** [comment]
It looks like the STR tree speeds up the processing quite a lot. Here's a simple benchmark using the Tasmania planning unit data. 

```
# load package
library(prioritizr)

# load example planning unit data
data(tas_pu, package = "prioritizrdata")

# make boundary matrix using standard functionality
system.time(m1 <- boundary_matrix(tas_pu))
#>  user  system elapsed 
#>  4.125   0.027   4.153

# make boundary matrix using STR queries
system.time(m2 <- boundary_matrix(tas_pu, TRUE))
#>  user  system elapsed 
#>  0.264   0.000   0.264 

# verify that solutions are same
all(m1 == m2)
#> [1] TRUE
```

**Comment by jeffreyhanson** [comment]
Hmm, it looks like there's a bug that only manifests sometimes and not all times causing the builds to fail. I'll see if I can track it down tomorrow.
